// relay-compiler-language-typescript/lib/TypeScriptTypeTransformers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformInputType = exports.transformScalarType = void 0;
var ts = require("typescript");
function getInputObjectTypeIdentifier(schema, typeID) {
    return schema.getTypeString(typeID);
}
function transformScalarType(schema, type, state, objectProps) {
    if (schema.isNonNull(type)) {
        return transformNonNullableScalarType(schema, schema.getNullableType(type), state, objectProps);
    }
    else {
        return ts.factory.createUnionTypeNode([
            transformNonNullableScalarType(schema, type, state, objectProps),
            ts.factory.createLiteralTypeNode(ts.factory.createToken(ts.SyntaxKind.NullKeyword)),
        ]);
    }
}
exports.transformScalarType = transformScalarType;
function transformNonNullableScalarType(schema, type, state, objectProps) {
    if (schema.isList(type)) {
        return ts.factory.createTypeReferenceNode(ts.factory.createIdentifier("ReadonlyArray"), [
            transformScalarType(schema, schema.getListItemType(type), state, objectProps),
        ]);
    }
    else if (schema.isObject(type) ||
        schema.isUnion(type) ||
        schema.isInterface(type)) {
        return objectProps;
    }
    else if (schema.isScalar(type)) {
        return transformGraphQLScalarType(schema.getTypeString(type), state);
    }
    else if (schema.isEnum(type)) {
        return transformGraphQLEnumType(schema, schema.assertEnumType(type), state);
    }
    else {
        throw new Error("Could not convert from GraphQL type " + type.toString());
    }
}
function transformGraphQLScalarType(typeName, state) {
    var customType = state.customScalars[typeName];
    switch (customType || typeName) {
        case "ID":
        case "String":
        case "Url":
            return ts.factory.createKeywordTypeNode(ts.SyntaxKind.StringKeyword);
        case "Float":
        case "Int":
            return ts.factory.createKeywordTypeNode(ts.SyntaxKind.NumberKeyword);
        case "Boolean":
            return ts.factory.createKeywordTypeNode(ts.SyntaxKind.BooleanKeyword);
        default:
            return customType
                ? ts.factory.createTypeReferenceNode(customType, undefined)
                : ts.factory.createKeywordTypeNode(ts.SyntaxKind.UnknownKeyword);
    }
}
function transformGraphQLEnumType(schema, type, state) {
    state.usedEnums[schema.getTypeString(type)] = type;
    return ts.factory.createTypeReferenceNode(ts.factory.createIdentifier(schema.getTypeString(type)), []);
}
function transformInputType(schema, type, state) {
    if (schema.isNonNull(type)) {
        return transformNonNullableInputType(schema, schema.getNullableType(type), state);
    }
    else {
        return ts.factory.createUnionTypeNode([
            transformNonNullableInputType(schema, type, state),
            ts.factory.createLiteralTypeNode(ts.factory.createToken(ts.SyntaxKind.NullKeyword)),
        ]);
    }
}
exports.transformInputType = transformInputType;
function transformNonNullableInputType(schema, type, state) {
    if (schema.isList(type)) {
        return ts.factory.createTypeReferenceNode(ts.factory.createIdentifier("Array"), [transformInputType(schema, schema.getListItemType(type), state)]);
    }
    else if (schema.isScalar(type)) {
        return transformGraphQLScalarType(schema.getTypeString(type), state);
    }
    else if (schema.isEnum(type)) {
        return transformGraphQLEnumType(schema, schema.assertEnumType(type), state);
    }
    else if (schema.isInputObject(type)) {
        var typeIdentifier = getInputObjectTypeIdentifier(schema, type);
        if (state.generatedInputObjectTypes[typeIdentifier]) {
            return ts.factory.createTypeReferenceNode(ts.factory.createIdentifier(typeIdentifier), []);
        }
        state.generatedInputObjectTypes[typeIdentifier] = "pending";
        var fields = schema.getFields(schema.assertInputObjectType(type));
        var props = fields.map(function (fieldID) {
            var fieldType = schema.getFieldType(fieldID);
            var fieldName = schema.getFieldName(fieldID);
            var property = ts.factory.createPropertySignature(undefined, ts.factory.createIdentifier(fieldName), state.optionalInputFields.indexOf(fieldName) >= 0 ||
                !schema.isNonNull(fieldType)
                ? ts.factory.createToken(ts.SyntaxKind.QuestionToken)
                : undefined, transformInputType(schema, fieldType, state));
            return property;
        });
        state.generatedInputObjectTypes[typeIdentifier] = ts.factory.createTypeLiteralNode(props);
        return ts.factory.createTypeReferenceNode(ts.factory.createIdentifier(typeIdentifier), []);
    }
    else {
        throw new Error("Could not convert from GraphQL type " + type.toString());
    }
}
